Add Blog for Verifying Raw Pointer Arithmetic Operations (#53)
model-checking/kani-verifier-blog: _posts/2024-12-10-verifying-raw-pointer-arithmetic-operations.md

#### Problem statement:
AWS is working to ensure the safety of Rust’s unsafe constructs using formal verification and automated reasoning. The problem statement we selected for our team was to verify the safety of standard library code that handles pointer arithmetic operations using the Kani verifier (a formal verification tool). By this, we aimed to strengthen trust in Rust’s safety guarantees in large-scale, safety-critical systems. 

#### What is pointer arithmetic?
Pointer arithmetic operations include addition, subtraction and offset which deal with raw pointer manipulation and accessing specific memory locations.
Pointer arithmetic is commonly used in applications that require precise control over memory such as operating systems, embedded systems and performance critical systems. If these operations are implemented incorrectly, they can cause serious harm and issues such as out-of-bounds memory access and data corruption/crashes.

#### What has our team has done?
Our team utilized Kani, a formal verification tool to check the safety of raw pointer arithmetic in Rust. Our achievements are:

> -	We implemented and verification function contracts for 16 pointer operations, such as add(), sub(), and offset().

Reviewer: ```suggestion
-	We implemented and verified function contracts for 16 pointer operations, such as `add()`, `sub()`, and `offset()`.
```

2. **Verification of Usages**: The goal was to ensure that the contracts for these methods were sufficient and that the usage of pointer arithmetic in other functions adhered to the defined safety guarantees. Documentation and analysis of function behavior were essential in defining input space and verifying that the functions operated safely across all expected inputs.

### Function Analysis

The analysis began with a thorough review of the functions listed for verification. This involved examining their implementation, studying the official Rust documentation, and understanding their safety requirements. This process was instrumental in identifying potential sources of undefined behavior and clarifying their intended usage. These insights were critical for defining precise and robust safety contracts.

> Among the functions analyzed, two stood out as foundational—offset and offset_from—due to their central role in enabling other pointer operations: 

Reviewer: ```suggestion
Among the functions analyzed, two stood out as foundational—`offset` and `offset_from`—due to their central role in enabling other pointer operations: 
```

By initially focusing on the safety of `offset` and `offset_from`, the verification effort focused on the fundamental components that underpin many related functions. This approach made it easier to define and extend contracts and proofs to dependent functions.

### Function Contracts

The preconditions and postconditions for the functions were primarily derived from the safety requirements outlined in the official Rust [documentation](https://doc.rust-lang.org/std/primitive.pointer.html#method.offset). For example, the safety requirements stated in the documentation for the `offset` function are as follows:
1. The offset in bytes, `count * size_of::<T>()`, computed on mathematical integers (without “wrapping around”), must fit in an `isize`.
2. If the computed offset is non-zero, then `self` must be derived from a pointer to some allocated object, and the entire memory range between `self` and the `result` must be in bounds of that allocated object. In particular, this range must not “wrap around” the edge of the address space.

#### Preconditions
The above safety requirements lead to the following preconditions:

> 1. If T is a zero-sized type, i.e., `size_of::<T>() == 0`, then the computed offset (`count * size_of::<T>()`) will always be 0. Thus, both safety checks are inherently satisfied, and no additional validations are required.

Reviewer: ```suggestion
1. If `T` is a zero-sized type, i.e., `size_of::<T>() == 0`, then the computed offset (`count * size_of::<T>()`) will always be 0. Thus, both safety checks are inherently satisfied, and no additional validations are required.
```

#### Example Proof for the `offset` Function

```rust
#[kani::proof_for_contract(<*const u8>::offset)]
pub fn check_const_add_i8() {
    // 200 bytes are large enough to cover all pointee types used for testing
    const BUF_SIZE: usize = 200;
    let mut generator = kani::PointerGenerator::<BUF_SIZE>::new();
    let test_ptr: *const u8 = generator.any_in_bounds().ptr;
    let count: isize = kani::any();
    unsafe {
        test_ptr.offset(count);
    }
}
```

This harness validates the `offset` function for pointers of type `*const u8`. It ensures that the function adheres to the safety contracts defined (as given in the previous section). The `<*const T>::offset` function accepts two arguments: a pointer (`*const T`) and an offset (`isize`). The proof generates these non-deterministically as follows:
* The `count` variable, of type `isize`, has an `Arbitrary` trait implemented, enabling the generation of non-deterministic values using `kani::any()`.
* `kani::PointerGenerator` is used to create a pointer `test_ptr`, guaranteed to lie within the bounds of the allocated buffer.

The `offset` function is called in an unsafe block with the generated `test_ptr` and `count`.

> PointerGenerator can also create pointers with different allocation statuses, such as out-of-bounds, dangling, or null pointers. This was particularly useful in writing harnesses for `offset_from`, which require testing pointers with varied allocation statuses (see [here](https://github.com/model-checking/verify-rust-std/blob/main/library/core/src/ptr/const_ptr.rs)). 

Reviewer: ```suggestion
Kani's pointer generator can also create pointers with different allocation statuses, such as out-of-bounds, dangling, or null pointers. This was particularly useful in writing harnesses for `offset_from`, which require testing pointers with varied allocation statuses (see [here](https://github.com/model-checking/verify-rust-std/blob/main/library/core/src/ptr/const_ptr.rs)). 
```

- The vector's length decreases by one after the operation.
- The removed element matches the original element at the specified index.
- If the removed index is not the last, the index now contains the last element of the original vector.
- All other elements remain unaffected.

#### **`Option::as_slice`**

The `Option::as_slice` method converts an `Option` containing a collection into a slice. The proof would validate:
- The result is a valid slice if the `Option` contains a value.
- The length of the resulting slice matches the length of the contained collection.

#### **`VecDeque::swap`**

The `VecDeque::swap` method swaps two elements at specified indices in a `VecDeque`. The Kani proof verifies:
- The elements at the specified indices are swapped correctly.
- All other elements in the `VecDeque` remain unchanged.

#### Summary of Usage Proofs

These proofs ensure the following:
- For `Vec::swap_remove`, the vector maintains integrity during element removals and replacements.
- For `Option::as_slice`, the resulting slice is valid and matches the contained data.
- For `VecDeque::swap`, swapped and unaffected elements are verified for correctness.

## Challenges

> Throughout the project, we faced several technical challenges related to Rust's specification, the Kani verifier's limitations, and the complex nature of pointer arithmetic operations. Below are some of the key challenges and how we addressed them or how the could be addressed in the future.

Reviewer: ```suggestion
Throughout the project, we faced several technical challenges related to Rust's specification, Kani's limitations, and the complex nature of pointer arithmetic operations. Below are some of the key challenges and how we addressed them or how the could be addressed in the future.
```
